sp/src/public/vscript/ivscript.h

```diff
@@ -739,6 +769,10 @@ static inline int ToConstantVariant(int value)
 #define DEFINE_SCRIPT_INSTANCE_HELPER( p )													pDesc->pHelper = (p);
 
 #ifdef MAPBASE_VSCRIPT
+// Allow instance to be deleted but not constructed
+// Not needed if the class has a constructor
+#define DEFINE_SCRIPT_REFCOUNTED_INSTANCE()													do { pDesc->m_pfnDestruct = &CScriptConstructor<_className>::Destruct; } while (0);
```

Comment:
`DEFINE_SCRIPT_REFCOUNTED_INSTANCE` is mirroring `DEFINE_SCRIPT_CONSTRUCTOR`. The name `REFCOUNTED_INSTANCE` is about intent instead of internals. Someone looking to register a C++ class doesn't need to know how it works. This could also be added to classes with constructors and the duplicated assignment would likely be optimised out.

> Having no construction, only destruction, is not really specific to ref-counted objects. Anything that vscript cannot create itself, but manages some state that needs to be cleaned up via a destruction-function would use this.

What are the examples of this?

> It would also be perfectly reasonable to have a ref-counted object that can be created from vscript.

This is what constructors do, that is the case in `FireBulletsInfo_t`. It can be both constructed in script and returned from `CreateFireBulletsInfo`.

Reply:
> The name `REFCOUNTED_INSTANCE` is about intent instead of internals. Someone looking to register a C++ class doesn't need to know how it works.

I may have missed something here. I was under the impression that this refers to reference-counting in C++, like the `CScriptKeyValues` do, though it may be intertwined with the ref-counting done through the SquirrelVM? Doesn't every script-instance do reference-counting though, (the reference-counting portion of) which is completely detached to destructors/Destruct-functions?

> What are the examples of this?

Well, this is a new implementation. I don't really have any new things in mind for which I'd like to use this.
There are existing classes that you're converting though that don't have intrinsic ref-counting, like the `FireBulletsInfo_t` you mention. Or does that happen somewhere behind the scenes, or did you intend to refer to the SquirrelVM reference-counting as I wonder about above?

Reply:
All non-primitive (not int, bool, float) script objects always do reference counting in Squirrel. `DEFINE_SCRIPT_REFCOUNTED_INSTANCE` and `HSCRIPT_RC` are for taking advantage of that. `CScriptKeyValues` is a kind of its own because it can create children that have memory that depend on the root, so it manually does reference counting of `KeyValues` in C++ independent of Squirrel, which only reference count for `CScriptKeyValues`.

Destructors need to be defined for script refcounted objects so that C++ can allocate memory, pass it to `RegisterInstance` then forget it all. Otherwise both the script object (`HSCRIPT`) and the memory it points to (`T *pInstance`) needs to be freed in C++ (see `CBaseEntity`).

I tried to explain this and how script objects work in the long reply below.